### ISSUE TITLE
protecting resources under WEB-INF

### ISSUE THREAD

**Issue by ren-zhijun-oracle** [eclipse-ee4j/mojarra#202]
The servlet specification states the following (SRV.9.5)

No file contained in the WEB-INF directory may
be served directly to a client by the container.

So if I have page "foo.jsp" saved in "WEB-INF/foo.jsp", I will
get a 404.

If I use Faces, I would assume that the same behavior should be enforced.

If I define the JSF front controller using an extension mapping as follows:

<servlet-mapping>
<servlet-name>Faces Servlet</servlet-name>
<url-pattern>*.jsf</url-pattern>
</servlet-mapping>

then everything works fine. All JSP files located under WEB-INF/ are
protected.

However, if I use prefix mapping as follows:

<servlet-mapping>
<servlet-name>Faces Servlet</servlet-name>
<url-pattern>/faces/*</url-pattern>
</servlet-mapping>

Then all JSP files located under WEB-INF/ are directly accessible,
breaking the contract defined by the JSP spec.

I assume that the reason why prefix mapping does not behave properly
is because the Faces front controller does not protect WEB-INF/ since
this is normally handled by the servlet container. The servlet container
does its job when the extension mapping is used because /WEB-INF is the
first token following the context name. When prefix mapping is used,
the servlet container does not have anything to protect since the first
token after the context name is the prefix 'faces'.

To be able to protect JSP pages located under WEB-INF/ while using
prefix mapping, one must add a security constraint as follows:

<security-constraint>
<web-resource-collection>
<url-pattern>/faces/WEB-INF/*</url-pattern>
</web-resource-collection>
<auth-constraint>
<description>
With no roles defined, no access granted
</description>
</auth-constraint>
</security-constraint>

This should not be required.

In summary, whether one uses prefix mapping or extension mapping,
Faces should be consistent and abide by the servlet spec which specifies
that resources located under WEB-INF cannot be served directly to
a client by the container.
#### Environment
Operating System: All
Platform: All
#### Affected Versions
[1.1]
